include/gatekeeper_gk.h

```diff
@@ -24,13 +24,116 @@
 #include "gatekeeper_ipip.h"
 #include "gatekeeper_ggu.h"
 #include "gatekeeper_mailbox.h"
+#include "gatekeeper_lpm.h"
+
+/* The maximum number of gateways that Gatekeeper may use. */
+#define GK_MAX_NUM_GW       (8)
+
+/*
+ * The LPM supports 8-bit next hop,
+ * which at most has 128 different values.
+ */
+#define GK_MAX_NUM_POLICIES (128)
 
 /*
  * A flow entry can be in one of three states:
  * request, granted, or declined.
  */
 enum gk_flow_state { GK_REQUEST, GK_GRANTED, GK_DECLINED };
 
+enum gk_policy_action {
+	/* Forward the packe to the corresponding Grantor. */
+	GK_FWD_GRANTOR,
+
+	/* Forward the packe to the corresponding gateway. */
+	GK_FWD_GATEWAY,
+
+	/*
+	 * The Grantor server is a neighbor.
+	 * Forward the packe to the Grantor server directly.
+	 */
+	GK_FWD_NEIGHBOR,
+
+	/* Forward the packet to the back interface. */
+	GK_FWD_BACK_NET,
+
+	/* Drop the packet. */
+	GK_DROP,
+
+	/* Invalid policy action. */
+	GK_POLICY_MAX,
+};
+
+struct gk_nexthop {
+
+	uint16_t proto;
+
+	union {
+		struct in_addr  v4;
+		struct in6_addr v6;
+	} ip;
+
+	/* Indicate whether the MAC address is stale or not. */
+	bool stale;
+
+	/* Cache the whole Ethernet header. */
+	struct ether_hdr eth_hdr;
+};
+
+struct gk_policy {
+
+	/* The policy action. */
+	enum gk_policy_action   action;
+
+	/*
+	 * The count of how many time the LPM tables refer to it,
+	 * so a policy can go away only when no LPM entry referring to it.
+	 */
+	uint32_t                ref_cnt;
+
+	union {
+		struct {
+			/*
+	 		 * The gateway for the grantor when
+	 		 * the Grantor server is not in the back network.
+	 		 */
+			struct gk_nexthop *nh;
+
+			/* The Grantor tunnel information. */
+			struct ipip_tunnel_info tunnel;
+		} gateway;
+
+		struct {
+			/* The Grantor server is treated as a gateway. */
+			struct gk_nexthop nh;
+
+			/*
+		 	 * When the Grantor is in the back network, no need
+			 * for gateway information, and only needs tunnel
+			 * information.
+		 	 */
+			struct ipip_tunnel_info tunnel;
+		} neigh;
```

Comment:
When the Grantor server is in the back network, that is for action ``GK_FWD_NEIGH``, where should we put the ``struct ether_hdr`` information? Can we combine the ``struct gateway`` and ``struct neigh`` into ``struct nexthop``?